schemas/mobility.schema

```diff
@@ -22,4 +22,16 @@ n_trips_250_500 	   INTEGER,
 n_trips_gte_500 	   INTEGER
 );
 
+CREATE TABLE $PG_SCHEMA_MOBILITY.airtravel (
+ts  				   TIMESTAMP NOT NULL,
+origin_code   		   TEXT,
+origin_city   		   TEXT,
```

Comment:
Have you guys checked if the referenced `main.locale` table contains the city? If it does, we might want to eliminate duplication here. If it doesn't, probably better to keep this here.

Reply:
I tried to merge these using the `origin_fips` and `dest_fips` to the `main.locale` table but it was unfortunately extremely slow. I do think that removing the duplication of information would be ideal, but it will slow down the ingest fairly substantially.

Reply:
I was under an impression we actually had most of the cities in `main.locale`. After checking I now know that impression was wrong. In fact, it appears we don't have any cities in the database. I suppose that's not surprising. Still, I think linking things would be good to have if possible. Do you mind if I try to get a better idea of why this is slow before we go without foreign keys?

Because we don't have cities, we'd need to go with counties. I doubt this would introduce too much noise given that most of the population of a county with a large city in it lives in that city. Am I correct assuming this is what you've tried to do with `origin_fips` and `dest_fips`? If yes, was it because `main.locale` lookup was slow? Do you think adding a full-text index on `main.locale.fips` column would help? The FIPS code is not really a number in that it has meaningful leading zeros (hence its data type is `text`), but we could add a `main.locale.fips_int` integer column with a B-TREE index (or HASH index if it would turn out to be much faster) that might speed things up. Or was is slow because there simply were too many origin and destination DB lookups? If that was the case, do you think using a Python dict (which would be created once only) for FIPS-to-locale_id lookup would help here at all?

Using foreign-key constraints (deferred or not) will trigger data integrity checks and that will make the inject transaction run slower; no way around that. The question is, is it worth paying that bigger price once or going with a bunch of smaller ones paid every time data is accessed.

Reply:
That's exactly right, the `origin_fips` and `dest_fips` are for the county where the airport is. 

I can try to run some experiments next week with an index and/or creating `fips_int` columns since I agree that having foreign keys would be far preferable! 

I'll also benchmark the speed on my machine and let you know just how long we're talking with the various options.

Reply:
@ubiq-x I've just pushed a commit to this PR that adds joins on the `main.locale` table for the origin and destination for each trip segment.

The current logic is such that when a user specifies a state to obtain air traffic mobility data (e..g `TX`) the data is filtered so that they get trips where the destination was in Texas. So, the lookup for destinations is easy (just a join on `fips`).

The origins may be from all over the world. So there are 3 lookups:

1. join on `fips` which covers US origin cities
2. join on `admin1` which covers origins that have `admin1` data spelled the same between BTS and the `locale` table
3. join on `admin0` which fills in the gaps left and just matches the origin city with that country's `locale id`. 

For context, the 2019 data for Texas had 1,760 records (origin, destination, month combinations) and only 142 of those end up with `null` values for `origin_locale` (they fail to match any of 1, 2 or 3 above).

Additionally, one thing to note is that the `mobility.schema` now has some redundancy with the `main.locale` table schema since it includes some fields which are used to perform the join. These could be dropped after the joins are finished, but then the `mobility.schema` file would not reflect what is actually in the database, which seems suboptimal. 

Welcome thoughts!

Reply:
FYI it takes just a minute or so to run a state for 1 year locally for me now.

Reply:
Thanks, @brandomr. That looks great!  I think the 1,2,3 approach you've taken makes a lot of sense.

Yes, I see the redundancy you've mentioned.  One thing that comes to mind and which would satisfy normalization is to add the missing locales to `main.locale` table when they're discovered and link to them.  And perhaps add a `main.locale.comment:TEXT` column to mark those locales as "added for mobility.airtraffic" or something.  That would lead to removing the following six columns from `mobility.airtraffic` (`*_code` and `*_city` would need to stay):

`origin_admin0`
`origin_admin1`
`origin_fips`
`dest_admin0`
`dest_admin1`
`dest_fips`

That's also some space savings albeit not sure how significant that would end up being.  I suppose that might add up given these are strings...  It would be roughly 10kB for `admin1` for Texas 2019 (1,760*6B) so probably several MB for all the states per year.  Not sure how much savings exactly that would end up being.  We could just roll with some redundancy.  Besides, if it started to bite us in the ass at some point, it shouldn't be hard to quick-fix it then.  What do you think?

Also, a minute is fast!  Hawk eyes points for spotting and changing the `airtraffic` table to unlogged.  I often forget about that myself.

Reply:
So I think one thing we could do/try is to add those 6 columns in the `localedb_man.py` "manually" and keep them out of the schema file. Then delete them once we are done using them.

Reply:
Good question; haven't thought of how to actually implement this, but the following seem like reasonable candidates:

1. Leave the columns in the `airtraffic` table for loading, then drop them, followed by `VACUUM FULL mobility.airtraffic;` to reclaim the space.
2. Use a temporary table.
3. Don't have the columns in the schema in the first place because if a locale doesn't exist in `main.locale` it would get added prior to the corresponding `mobility.airtraffic` record.

There's probably other options and it would depend on how you the actual loading set up.  If I read your code correctly, I think option #3 looks the least palleteable.

Reply:
I've just realized I misread your comment.  Yes, I suppose in terms of the schema file itself, it'd probably be best to not have those columns there.  Manually adding those columns is probably the only option.  Or having a separate "load schema."

Reply:
@ubiq-x the latest two commits I pushed resolve this:

First, temporary columns are added to support the required joins anytime new airtraffic data is added. 

If any locales are `null` (e.g. they don't exist in `main.locale`) they are added there. Their `id` in that table is set to a random integer between `9000000` and `9999999` though this could be changed if you'd like. Unfortunately, for some of these locales we just have `admin0` and nothing more ☹️ 

The corresponding rows in `mobility.airtraffic` are updated to reflect these _"newly discovered"_ locales.

Then, the temporary columns are removed.

Reply:
@brandomr awesome! I like the idea of an integer range dedicated to the new locales. Do you think there's a chance of an ID conflict given they are random?

Reply:
I looked through the current `ids` and it seemed like they were a combination of ISO country codes and/or FIPS. We could add a couple digits to the random `id` to make it even larger and reduce the likelihood of collision? Or pick a different range for them? 

I _think_ that since `id` is a primary key and I'm adding these to `main.locale` as an `insert` it would error out if if hits a collision.

Thoughts?

Reply:
Oh, yes, I'm pretty sure there won't be any conflicts with the "original" content; the IDs are all below the lower bound of the random sampling interval.  As I look at the code, `rd.randint()` may cause "self-conflict" with the air-traffic locales.  The sampling interval is `999999` wide so the probability of that happening is minuscule, but it's not zero.  How about switching to sequential IDs?

Reply:
Just converted `main.locale` `id` column to `SERIAL` in the latest commit and fixed the logic around adding locales to account for that (no need to specify `id` now since it is auto-incremented). That removes the possibility of any future collisions since `id` is never user defined.

Reply:
Good idea! I originally kept it as `INTEGER` because those values were being pulled directly from the JHU's locales dataset. It's more elegant now.